File: ioi/2010/maze_sa.py
#!/usr/bin/env python3
"""
Simulated annealing for IOI 2010 Maze (output-only), following the Codeforces
methodology: random # <-> . flips, energy = max BFS distance (in cells) from
the unique perimeter '.', Boltzmann acceptance, geometric cooling.

IOI statement constraints (stricter than the original CF snippet):
  * Obstacles X stay fixed.
  * Exactly one '.' on the outer border (entrance); every other '.' lies in
    the strict interior (rows/cols not on the field edge).

Neighbour moves:
  * Interior corn cell: toggle '#' <-> '.'.
  * Perimeter corn cell: relocate the entrance (old entrance '#', chosen cell '.'),
    with the same corner case as the reference code when the sampled cell is the
    current entrance (pick another random side).

calc_energy: multi-source BFS from entrance — maximum shortest-path length in
cells (entrance counts as 1), matching ``grade_maze`` / ``eccentricity_energy``.
"""

from __future__ import annotations

import argparse
import math
import random
import sys
import time
from collections import deque
from pathlib import Path
from typing import List, Optional, Sequence, Tuple

Coord = Tuple[int, int]

DIRS = ((1, 0), (-1, 0), (0, 1), (0, -1))


def parse_grid(lines: List[str]) -> Tuple[List[str], int, int]:
    g = [ln.rstrip("\r\n") for ln in lines if ln.strip() != ""]
    n, m = len(g), len(g[0])
    return g, n, m


def is_perimeter(i: int, j: int, n: int, m: int) -> bool:
    return i == 0 or i == n - 1 or j == 0 or j == m - 1


def strict_interior(i: int, j: int, n: int, m: int) -> bool:
    return 1 <= i <= n - 2 and 1 <= j <= m - 2


def is_corner(i: int, j: int, n: int, m: int) -> bool:
    return (i == 0 and j == 0) or (i == 0 and j == m - 1) or (
        i == n - 1 and j == 0
    ) or (i == n - 1 and j == m - 1)


def build_sa_masks(
    g_in: Sequence[str], n: int, m: int
) -> Tuple[List[Coord], List[Coord], List[Coord]]:
    """
    CF blog (entry 118853): ``sides`` = perimeter corn excluding corners;
    ``all`` = every non-corner corn cell (interior + boundary).
    """
    sides: List[Coord] = []
    all_cells: List[Coord] = []
    for i in range(1, n - 1):
        for j in (0, m - 1):
            if g_in[i][j] == "#":
                sides.append((i, j))
    for j in range(1, m - 1):
        for i in (0, n - 1):
            if g_in[i][j] == "#":
                sides.append((i, j))
    for i in range(n):
        for j in range(m):
            if is_corner(i, j, n, m) or g_in[i][j] != "#":
                continue
            all_cells.append((i, j))
    interior = [c for c in all_cells if strict_interior(c[0], c[1], n, m)]
    return sides, interior, all_cells


def random_initial_state(
    g_in: List[str],
    n: int,
    m: int,
    sides: Sequence[Coord],
    interior: Sequence[Coord],
    rng: random.Random,
    interior_dot_prob: float = 0.7,
) -> Tuple[List[List[str]], Coord]:
    """Random entrance on perimeter + random interior '.' fraction (blog-style)."""
    if not sides:
        raise ValueError("no valid entrance (perimeter corn)")
    src = rng.choice(sides)
    g = [list(row) for row in g_in]
    for i, j in interior:
        if rng.random() < interior_dot_prob:
            g[i][j] = "."
        else:
            g[i][j] = "#"
    er, ec = src
    g[er][ec] = "."
    return g, src


def load_state_from_output(
    g_in: List[str],
    path: Path,
) -> Optional[Tuple[List[List[str]], Coord]]:
    """Load maze output; return None if invalid."""
    text = path.read_text(encoding="utf-8")
    lines = [ln.rstrip("\n") for ln in text.splitlines() if ln.strip() != ""]
    if len(lines) != len(g_in):
        return None
    n, m = len(g_in), len(g_in[0])
    g = [list(row) for row in lines]
    ent: Optional[Coord] = None
    for i in range(n):
        if len(g[i]) != m:
            return None
        for j in range(m):
            if g_in[i][j] == "X" and g[i][j] != "X":
                return None
            if g_in[i][j] == "#":
                if g[i][j] not in "#.":
                    return None
            elif g_in[i][j] == "X":
                if g[i][j] != "X":
                    return None
            if g[i][j] == "." and is_perimeter(i, j, n, m):
                if ent is not None:
                    return None
                ent = (i, j)
    if ent is None:
        return None
    return g, ent


def calc_energy(g: List[List[str]], n: int, m: int, src: Coord) -> int:
    """Max BFS distance from entrance (cells), or 0 if entrance missing."""
    si, sj = src
    if g[si][sj] != ".":
        return 0
    dist = [[-1] * m for _ in range(n)]
    dist[si][sj] = 1
    q: deque[Coord] = deque([(si, sj)])
    best = 1
    while q:
        i, j = q.popleft()
        d0 = dist[i][j]
        best = max(best, d0)
        for di, dj in DIRS:
            ni, nj = i + di, j + dj
            if not (0 <= ni < n and 0 <= nj < m):
                continue
            if g[ni][nj] != ".":
                continue
            if dist[ni][nj] != -1:
                continue
            dist[ni][nj] = d0 + 1
            q.append((ni, nj))
    return best


def grid_to_lines(g: List[List[str]]) -> List[str]:
    return ["".join(row) for row in g]


def run_sa(
    g_in: List[str],
    rng: random.Random,
    t0: float,
    alpha: float,
    max_seconds: Optional[float],
    max_iters: Optional[int],
    save_every: int,
    out_path: Optional[Path],
    warm_path: Optional[Path],
    log_every: int,
) -> Tuple[List[str], int]:
    n, m = len(g_in), len(g_in[0])
    sides, interior, all_cells = build_sa_masks(g_in, n, m)
    if not sides:
        raise SystemExit("No perimeter corn — cannot place entrance.")

    flat_pick: List[Coord] = all_cells

    if warm_path and warm_path.is_file():
        loaded = load_state_from_output(g_in, warm_path)
        if loaded:
            g, src = loaded
            print(f"Warm start from {warm_path}", file=sys.stderr)
        else:
            g, src = random_initial_state(g_in, n, m, sides, interior, rng)
            print("Warm start invalid; using random initial state", file=sys.stderr)
    else:
        g, src = random_initial_state(g_in, n, m, sides, interior, rng)

    src_ref: List[Coord] = [src]
    cur_e = calc_energy(g, n, m, src_ref[0])
    best_e = cur_e
    best_grid = [row[:] for row in g]
    best_src = src_ref[0]

    t = t0
    it = 0
    t_start = time.monotonic()

    while True:
        if max_iters is not None and it >= max_iters:
            break
        if max_seconds is not None and (time.monotonic() - t_start) >= max_seconds:
            break
        it += 1
        old_src = src_ref[0]
        i, j = rng.choice(flat_pick)

        if not is_perimeter(i, j, n, m):
            old_ch = g[i][j]
            g[i][j] = "#" if old_ch == "." else "."
        else:
            g[old_src[0]][old_src[1]] = "#"
            if old_src == (i, j):
                src_ref[0] = rng.choice(sides)
            else:
                src_ref[0] = (i, j)
            g[src_ref[0][0]][src_ref[0][1]] = "."

        new_e = calc_energy(g, n, m, src_ref[0])

        if new_e >= best_e:
            best_e = new_e
            best_grid = [row[:] for row in g]
            best_src = src_ref[0]

        accept = new_e >= cur_e or (
            t > 1e-300 and rng.random() < math.exp((new_e - cur_e) / t)
        )

        if accept:
            cur_e = new_e
        else:
            if not is_perimeter(i, j, n, m):
                g[i][j] = old_ch
            else:
                g[src_ref[0][0]][src_ref[0][1]] = "#"
                src_ref[0] = old_src
                g[old_src[0]][old_src[1]] = "."

        t *= alpha

        if log_every > 0 and it % log_every == 0:
            elapsed = time.monotonic() - t_start
            print(
                f"iter={it}  t={t:.6g}  curE={cur_e}  bestE={best_e}  elapsed={elapsed:.1f}s",
                file=sys.stderr,
            )

        if save_every > 0 and out_path and it % save_every == 0:
            _write_output(out_path, best_grid, g_in, n, m, best_src)

    # final write uses best maze
    lines = _finalize_lines(best_grid, g_in, n, m, best_src)
    if out_path:
        out_path.write_text("\n".join(lines) + "\n", encoding="utf-8")
    return lines, best_e


def _write_output(
    out: Path,
    g: List[List[str]],
    g_in: List[str],
    n: int,
    m: int,
    src: Coord,
) -> None:
    lines = _finalize_lines(g, g_in, n, m, src)
    out.write_text("\n".join(lines) + "\n", encoding="utf-8")


def _finalize_lines(
    g: List[List[str]],
    g_in: List[str],
    n: int,
    m: int,
    src: Coord,
) -> List[str]:
    """Ensure X fixed, corn only #/., single entrance."""
    rows = [list(row) for row in g]
    for i in range(n):
        for j in range(m):
            if g_in[i][j] == "X":
                rows[i][j] = "X"
            elif g_in[i][j] == "#":
                if rows[i][j] not in "#.":
                    rows[i][j] = "#"
    er, ec = src
    for i in range(n):
        for j in range(m):
            if g_in[i][j] != "#":
                continue
            if is_perimeter(i, j, n, m) and (i, j) != (er, ec):
                rows[i][j] = "#"
    rows[er][ec] = "."
    return ["".join(r) for r in rows]


def main() -> None:
    ap = argparse.ArgumentParser(
        description="IOI 2010 Maze — simulated annealing (Boltzmann + geometric cooling).",
        epilog=(
            "Example long run:  python maze_sa.py g_inputs/tests/02 -o 02.out "
            "--alpha 0.999999999 --max-seconds 864000 --warm maze_outputs/02.out "
            "--save-every 1000000 --log-every 500000"
        ),
    )
    ap.add_argument("input", type=Path)
    ap.add_argument("-o", "--output", type=Path, required=True)
    ap.add_argument("--seed", type=int, default=0)
    ap.add_argument("--t0", type=float, default=2.5)
    ap.add_argument("--alpha", type=float, default=0.999999999)
    ap.add_argument("--max-seconds", type=float, default=None)
    ap.add_argument("--max-iters", type=int, default=None)
    ap.add_argument(
        "--save-every",
        type=int,
        default=1 << 23,
        help="write best grid every N iterations (blog uses 2^23)",
    )
    ap.add_argument("--warm", type=Path, default=None, help="optional starting .out")
    ap.add_argument("--log-every", type=int, default=0)
    args = ap.parse_args()

    text = args.input.read_text(encoding="utf-8")
    g_in, _, _ = parse_grid(text.splitlines())

    rng = random.Random(args.seed)
    lines, best_e = run_sa(
        g_in,
        rng,
        t0=args.t0,
        alpha=args.alpha,
        max_seconds=args.max_seconds,
        max_iters=args.max_iters,
        save_every=args.save_every,
        out_path=args.output,
        warm_path=args.warm,
        log_every=args.log_every,
    )
    print(f"best_energy={best_e}  -> {args.output}", file=sys.stderr)


if __name__ == "__main__":
    main()

File: ioi/2010/maze_sa.cc
// IOI 2010 Maze — simulated annealing (maomao90 / CF blog entry 118853).
//
// Neighbour: pick a random corn cell (non-corner, non-X); toggle # <-> . on the
// interior, or relocate the unique perimeter entrance on the boundary.
// Energy: max BFS distance in cells from the entrance ('.' on the border).
// Cooling: t0 = 2.5, alpha = 0.999999999, Boltzmann acceptance.
//
// Usage:  maze_sa <input> <output> [max_iters] [--seed N] [--warm path]
//   max_iters omitted => run until killed (blog style; saves best every 2^23 steps).

#include <bits/stdc++.h>
using namespace std;

namespace {

constexpr int k_max_n = 205;
constexpr int k_max_cells = k_max_n * k_max_n;
constexpr double k_t0 = 2.5;
constexpr double k_alpha = 0.999999999;
constexpr int k_save_mask = (1 << 23) - 1;

const int k_di[4] = {1, 0, -1, 0};
const int k_dj[4] = {0, 1, 0, -1};

int n, m;
char g[k_max_n][k_max_n];
char bestg[k_max_n][k_max_n];
char input_mask[k_max_n][k_max_n];

int dist[k_max_n][k_max_n];
unsigned short visit_stamp[k_max_n][k_max_n];
unsigned short visit_gen = 1;

int bfs_q_r[k_max_cells];
int bfs_q_c[k_max_cells];

vector<pair<int, int>> sides;
vector<pair<int, int>> all;
pair<int, int> src = {-1, -1};
pair<int, int> best_src = {-1, -1};

mt19937 rng(0);
uniform_int_distribution<int> pick_all;
uniform_int_distribution<int> pick_side;
uniform_real_distribution<double> pick_real;

bool is_corner(int i, int j) {
  return (i == 0 && j == 0) || (i == 0 && j == m - 1) || (i == n - 1 && j == 0) ||
         (i == n - 1 && j == m - 1);
}

bool is_perimeter(int i, int j) {
  return i == 0 || i == n - 1 || j == 0 || j == m - 1;
}

void build_masks() {
  sides.clear();
  all.clear();

  for (int i = 1; i < n - 1; ++i) {
    for (int j : {0, m - 1}) {
      if (input_mask[i][j] == '#') {
        sides.emplace_back(i, j);
      }
    }
  }
  for (int j = 1; j < m - 1; ++j) {
    for (int i : {0, n - 1}) {
      if (input_mask[i][j] == '#') {
        sides.emplace_back(i, j);
      }
    }
  }

  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < m; ++j) {
      if (is_corner(i, j)) {
        continue;
      }
      if (input_mask[i][j] != '#') {
        continue;
      }
      all.emplace_back(i, j);
    }
  }

  pick_all = uniform_int_distribution<int>(0, static_cast<int>(all.size()) - 1);
  pick_side = uniform_int_distribution<int>(0, static_cast<int>(sides.size()) - 1);
}

int calc_energy() {
  ++visit_gen;
  if (visit_gen == 0) {
    memset(visit_stamp, 0, sizeof visit_stamp);
    visit_gen = 1;
  }

  const int si = src.first;
  const int sj = src.second;
  if (g[si][sj] != '.') {
    return 0;
  }

  int head = 0;
  int tail = 0;
  bfs_q_r[tail] = si;
  bfs_q_c[tail] = sj;
  ++tail;
  visit_stamp[si][sj] = visit_gen;
  dist[si][sj] = 1;
  int res = 1;

  while (head < tail) {
    const int i = bfs_q_r[head];
    const int j = bfs_q_c[head];
    ++head;
    const int d0 = dist[i][j];
    res = max(res, d0);
    for (int k = 0; k < 4; ++k) {
      const int ni = i + k_di[k];
      const int nj = j + k_dj[k];
      if (ni < 0 || ni >= n || nj < 0 || nj >= m) {
        continue;
      }
      if (g[ni][nj] != '.') {
        continue;
      }
      if (visit_stamp[ni][nj] == visit_gen) {
        continue;
      }
      visit_stamp[ni][nj] = visit_gen;
      dist[ni][nj] = d0 + 1;
      bfs_q_r[tail] = ni;
      bfs_q_c[tail] = nj;
      ++tail;
    }
  }
  return res;
}

bool accept(int new_energy, int old_energy, double t) {
  if (new_energy >= old_energy) {
    return true;
  }
  return pick_real(rng) < exp((new_energy - old_energy) / t);
}

void write_grid(const char grid[k_max_n][k_max_n], const pair<int, int>& entrance,
                const char* path) {
  ofstream fout(path);
  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < m; ++j) {
      if (input_mask[i][j] == 'X') {
        fout << 'X';
        continue;
      }
      if (input_mask[i][j] != '#') {
        continue;
      }
      char ch = grid[i][j];
      if (is_perimeter(i, j)) {
        const bool is_ent = (i == entrance.first && j == entrance.second);
        ch = is_ent ? '.' : '#';
      }
      fout << ch;
    }
    fout << '\n';
  }
}

void random_initial_state() {
  if (sides.empty()) {
    fprintf(stderr, "no valid entrance cells\n");
    exit(1);
  }

  src = sides[pick_side(rng)];
  g[src.first][src.second] = '.';

  uniform_int_distribution<int> coin(0, 9);
  for (int i = 1; i < n - 1; ++i) {
    for (int j = 1; j < m - 1; ++j) {
      if (input_mask[i][j] != '#') {
        continue;
      }
      g[i][j] = (coin(rng) < 7) ? '.' : '#';
    }
  }
}

void load_warm(const char* path) {
  ifstream fin(path);
  if (!fin) {
    return;
  }
  string line;
  int row = 0;
  while (getline(fin, line) && row < n) {
    for (int j = 0; j < m && j < static_cast<int>(line.size()); ++j) {
      g[row][j] = line[j];
      if (g[row][j] == '.' && is_perimeter(row, j)) {
        src = {row, j};
      }
    }
    ++row;
  }
}

bool valid_warm() {
  if (src.first < 0) {
    return false;
  }
  int perimeter_dots = 0;
  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < m; ++j) {
      if (input_mask[i][j] == 'X' && g[i][j] != 'X') {
        return false;
      }
      if (g[i][j] == '.' && is_perimeter(i, j)) {
        ++perimeter_dots;
      }
    }
  }
  return perimeter_dots == 1;
}

void read_input(const char* path) {
  ifstream fin(path);
  if (!fin) {
    perror(path);
    exit(1);
  }
  string line;
  while (getline(fin, line)) {
    while (!line.empty() && (line.back() == '\r' || line.back() == '\n')) {
      line.pop_back();
    }
    if (line.empty()) {
      continue;
    }
    if (n >= k_max_n) {
      break;
    }
    m = static_cast<int>(line.size());
    for (int j = 0; j < m; ++j) {
      input_mask[n][j] = line[j];
      g[n][j] = line[j];
    }
    ++n;
  }
}

void save_best() {
  for (int i = 0; i < n; ++i) {
    memcpy(bestg[i], g[i], m);
  }
}

} // namespace

int main(int argc, char** argv) {
  if (argc < 3) {
    fprintf(stderr, "usage: %s <input> <output> [max_iters] [--seed N] [--warm path]\n", argv[0]);
    return 1;
  }

  const char* in_path = argv[1];
  const char* out_path = argv[2];
  long long max_iters = -1;
  const char* warm_path = nullptr;
  unsigned seed =
      static_cast<unsigned>(chrono::high_resolution_clock::now().time_since_epoch().count());

  for (int i = 3; i < argc; ++i) {
    if (string(argv[i]) == "--warm" && i + 1 < argc) {
      warm_path = argv[++i];
    } else if (string(argv[i]) == "--seed" && i + 1 < argc) {
      seed = static_cast<unsigned>(strtoul(argv[++i], nullptr, 10));
    } else {
      max_iters = atoll(argv[i]);
    }
  }

  rng.seed(seed);
  read_input(in_path);
  build_masks();

  bool warmed = false;
  if (warm_path) {
    load_warm(warm_path);
    warmed = valid_warm();
  }
  if (!warmed) {
    for (int i = 0; i < n; ++i) {
      for (int j = 0; j < m; ++j) {
        g[i][j] = input_mask[i][j];
      }
    }
    src = {-1, -1};
    random_initial_state();
  }

  double t = k_t0;
  int cur_energy = calc_energy();
  int best_energy = cur_energy;
  best_src = src;
  save_best();

  unsigned int cnt = 0;
  const auto t_start = chrono::steady_clock::now();

  while (max_iters < 0 || static_cast<long long>(cnt) < max_iters) {
    if ((cnt & k_save_mask) == 0) {
      fprintf(stderr, "%.12g %d\n", t, cur_energy);
      write_grid(bestg, best_src, out_path);
    }
    ++cnt;

    const int id = pick_all(rng);
    const int ci = all[id].first;
    const int cj = all[id].second;
    const auto old_src = src;

    if (is_perimeter(ci, cj)) {
      g[old_src.first][old_src.second] = '#';
      if (old_src.first == ci && old_src.second == cj) {
        src = sides[pick_side(rng)];
      } else {
        src = {ci, cj};
      }
      g[src.first][src.second] = '.';
    } else {
      g[ci][cj] = (g[ci][cj] == '#') ? '.' : '#';
    }

    const int new_energy = calc_energy();
    if (new_energy >= best_energy) {
      best_energy = new_energy;
      best_src = src;
      save_best();
    }

    if (accept(new_energy, cur_energy, t)) {
      cur_energy = new_energy;
    } else {
      if (is_perimeter(ci, cj)) {
        g[src.first][src.second] = '#';
        g[old_src.first][old_src.second] = '.';
        src = old_src;
      } else {
        g[ci][cj] = (g[ci][cj] == '#') ? '.' : '#';
      }
    }

    t *= k_alpha;
  }

  write_grid(bestg, best_src, out_path);
  const auto elapsed =
      chrono::duration_cast<chrono::milliseconds>(chrono::steady_clock::now() - t_start);
  fprintf(stderr, "best_energy=%d  iters=%u  elapsed_ms=%lld  seed=%u  -> %s\n", best_energy, cnt,
          static_cast<long long>(elapsed.count()), seed, out_path);
  return 0;
}
